File: part1/src/main/java/com/peterlaurence/book/javatokotlin/callbackpattern/fragments/java/PurchasesViewModel.java
package com.peterlaurence.book.javatokotlin.callbackpattern.fragments.java;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.peterlaurence.book.javatokotlin.callbackpattern.core.java.UserPurchases;

import java.util.List;

public class PurchasesViewModel extends ViewModel {
    private BillingClient billingClient;
    private final String user;

    private MutableLiveData<UserPurchases> purchases;

    PurchasesViewModel(BillingClient billingClient,
                       String user) {
        this.billingClient = billingClient;
        this.user = user;
    }

    private void getUserPurchases(String user) {
        billingClient.init(provider -> {
            // this is called from a background thread
            if (provider != null) {
                provider.fetchPurchases(user, purchases -> {
                    this.purchases.postValue(new UserPurchases(user, purchases));
                });
            }
        });
    }

    LiveData<UserPurchases> getPurchasesLiveData() {
        if (purchases == null) {
            purchases = new MutableLiveData<>();
            getUserPurchases(user);
        }
        return purchases;
    }

    public interface BillingClient {
        interface BillingCallback {
            void onInitDone(@Nullable PurchasesProvider provider);
        }

        void init(BillingCallback callback);
    }

    public interface PurchasesProvider {
        interface PurchaseFetchCallback {
            void onPurchaseFetchDone(List<String> purchases);
        }

        void fetchPurchases(String user, PurchaseFetchCallback callback);
    }
}

File: part1/src/main/java/com/peterlaurence/book/javatokotlin/callbackpattern/fragments/PurchasesViewModel.kt
package com.peterlaurence.book.javatokotlin.callbackpattern.fragments

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.peterlaurence.book.javatokotlin.callbackpattern.core.UserPurchases
import com.peterlaurence.book.javatokotlin.callbackpattern.fragments.PurchasesViewModel.BillingClient.BillingCallback
import com.peterlaurence.book.javatokotlin.callbackpattern.fragments.PurchasesViewModel.PurchasesProvider.PurchaseFetchCallback

class PurchasesViewModel internal constructor(
    private val billingClient: BillingClient,
    private val user: String
) : ViewModel() {
    private var _purchases = MutableLiveData<UserPurchases>()

    private fun getUserPurchases(user: String) {
        billingClient.init { provider -> // this is called from a background thread
            provider?.fetchPurchases(user) { purchases ->
                _purchases.postValue(UserPurchases(user, purchases))
            }
        }
    }

    val purchasesLiveData: LiveData<UserPurchases>
        get() {
            getUserPurchases(user)
            return _purchases
        }

    interface BillingClient {
        fun interface BillingCallback {
            fun onInitDone(provider: PurchasesProvider?)
        }

        fun init(callback: BillingCallback)
    }

    interface PurchasesProvider {
        fun interface PurchaseFetchCallback {
            fun onPurchaseFetchDone(purchases: List<String>)
        }

        fun fetchPurchases(user: String, callback: PurchaseFetchCallback)
    }

}